pkg/receive/handler.go

```diff
@@ -482,6 +528,63 @@ func (h *Handler) receiveHTTP(w http.ResponseWriter, r *http.Request) {
 	h.writeSamplesTotal.WithLabelValues(strconv.Itoa(responseStatusCode), tenant).Observe(float64(totalSamples))
 }
 
+// isUnderLimit ensures that the current number of active series for a tenant does not exceed given limit.
+// It does so in a best-effort way, i.e, in case meta-monitoring is unreachable, it does not impose limits.
+// TODO(saswatamcode): Add capability to configure diff limits for diff tenants.
+func (h *Handler) isUnderLimit(ctx context.Context, tenant string, logger log.Logger) (bool, error) {
+	if h.options.MaxPerTenantLimit == 0 || h.options.MetaMonitoringUrl.Host == "" {
+		return true, nil
+	}
+
+	// Use specified HTTPConfig to make requests to meta-monitoring.
+	httpConfContentYaml, err := h.options.MetaMonitoringHttpClient.Content()
+	if err != nil {
+		return true, errors.Wrap(err, "getting http client config")
+	}
+
+	httpClientConfig, err := httpconfig.NewClientConfigFromYAML(httpConfContentYaml)
+	if err != nil {
+		return true, errors.Wrap(err, "parsing http config YAML")
+	}
+
+	httpClient, err := httpconfig.NewHTTPClient(*httpClientConfig, "thanos-receive")
+	if err != nil {
+		return true, errors.Wrap(err, "improper http client config")
+	}
+
+	c := promclient.NewWithTracingClient(logger, httpClient, httpconfig.ThanosUserAgent)
+
+	vectorRes, _, err := c.QueryInstant(ctx, h.options.MetaMonitoringUrl, h.options.MetaMonitoringLimitQuery, time.Now(), promclient.QueryOptions{})
+	if err != nil {
+		return true, errors.Wrap(err, "failed to query meta-monitoring")
+	}
+
+	level.Debug(logger).Log("msg", "successfully queried meta-monitoring", "vectors", len(vectorRes))
+
+	// In such limiting flow, we ingest the first remote write request
+	// and then check meta-monitoring metric to ascertain current active
+	// series. As such metric is updated in intervals, it is possible
+	// that Receive ingests more series than the limit, before detecting that
+	// a tenant has exceeded the set limits.
+	for _, e := range vectorRes {
+		for k, v := range e.Metric {
+			// Search for metric which has tenant label for a particular tenant.
+			if k == "tenant" && string(v) == tenant {
+				h.aboveLimit.WithLabelValues(tenant).Set(float64(e.Value) - float64(h.options.MaxPerTenantLimit))
+				if float64(e.Value) >= float64(h.options.MaxPerTenantLimit) {
+					level.Error(logger).Log("msg", "tenant above limit", "currentSeries", float64(e.Value))
```

Comment:
Would be nice to have the limit on this log line, so we don't need to check elsewhere for the limit.

Reply:
Yes, will add! 🙂

Reply:
Added in https://github.com/thanos-io/thanos/pull/5520/commits/b647b541e4202fd6c91967fad20f35f2a94a8363